agent/ecr/factory.go

```diff
@@ -56,9 +58,22 @@ func NewECRFactory(acceptInsecureCert bool) ECRFactory {
 	}
 }
 
-// GetClient returns the correct region- and endpoint-aware client
-func (factory *ecrFactory) GetClient(region, endpointOverride string) ECRClient {
-	key := cacheKey{region: region, endpointOverride: endpointOverride}
+// GetClient returns the correct region and endpoint aware client
+func (factory *ecrFactory) GetClient(authData *api.ECRAuthData) ECRClient {
+	cfg := aws.NewConfig().WithRegion(authData.Region).WithHTTPClient(factory.httpClient)
+	if authData.EndpointOverride != "" {
+		cfg.Region = aws.String(authData.EndpointOverride)
+	}
+
+	// Container has its own credentials to pull from ECR
+	if authData.PullCredentials != nil {
+		creds := credentials.NewStaticCredentials(authData.PullCredentials.AccessKeyID, authData.PullCredentials.SecretAccessKey, authData.PullCredentials.SessionToken)
+		cfg := cfg.WithCredentials(creds)
+		return factory.newClient(cfg)
+	}
+
+	// For containers pull using the default credentials, cache the ecr client
+	key := cacheKey{region: authData.Region, endpointOverride: authData.EndpointOverride}
 	client, ok := factory.clients[key]
```

Comment:
out of curiosity and this isnt directly part of the changes you made, but L77-87 seems like duplicate logic. except one locks the `clients` map. do you know why we do this?